File: perl_solutions_21-30/Euler_21.pl
use strict;
use warnings;
use Time::HiRes qw( clock );


sub amicable_pairs
{
    my $total = 0;
    for my $d1 (1..9999) {
        my $p1 = 0;
        my $p2 = 0;
        for my $d2 (1..($d1 / 2) + 1) {
            if ($d1 % $d2 == 0) {
                $p1 += $d2;
            }
        }
        for my $d2 (1..($p1 / 2) + 1) {
            if ($p1 % $d2 == 0) {
                $p2 += $d2;
            }
        } 
        if ($p2 == $d1 && $d1 != $p1) {
            $total += $d1;
        }
    }
    return $total;
}

my $start = clock();
my $answer = amicable_pairs();
my $stop = clock();
print "Answer: $answer\n";
printf "Time: %.5f\n", ($stop - $start);

File: c_solutions_21-30/Euler_21.c
// Calculate the sum of all the amicable pairs below 10000

#include <time.h>
#include <stdio.h>

#define LIMIT 10000


int main(int argc, char *argv[])
{
    clock_t start, stop;
    start = clock();

    int answer, p1, p2; 
    int i, j;

    for (i=1, answer=0, p1=0, p2=0; i < LIMIT; i++, p1=0, p2=0) {
        for (j=1; j < (i / 2) + 1; j++) 
            p1 += i % j == 0 ? j : 0;
        for (j=1; j < (p1 / 2) + 1; j++)
            p2 += p1 % j == 0 ? j : 0;
        answer += (p2 == i) && (i != p1) ? i : 0;
    }

    stop = clock();

    printf("Answer: %d\n", answer);
    printf("Time: %f\n", ((float) stop - (float) start) / CLOCKS_PER_SEC);

    return 0;
}
